fix(Table): improve `data` reactivity

zguig52 opened the pull request:
<!---
☝️ PR title should follow conventional commits (https://conventionalcommits.org)
-->

### 🔗 Linked issue
Resolves https://github.com/nuxt/ui/issues/3962 https://github.com/nuxt/ui/issues/3377 https://github.com/nuxt/ui/issues/3597 https://github.com/nuxt/ui/issues/3821
<!-- If it resolves an open issue, please link the issue here. For example "Resolves #123" -->

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [ ] 📖 Documentation (updates to the documentation or readme)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

<!-- Describe your changes in detail -->
<!-- Why is this change required? What problem does it solve? -->
Tables are no more reactive in v3, being based on TanStack Table.
By using an internal ref + watcher that automatically trigger data re-assignation to be detected/updated by TanStack.
Optionally, in case of performances issues, original behaviour can be restored by setting watchDepth prop to false. Depth can also be fine tuned to limit it.

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

zguig52 committed:
feat(Table): allow Table reactivity + associated watcher depth configuration (for performances optimization capabilities)

Restore Nuxt UI v2 behaviour of table:
https://github.com/nuxt/ui/issues/3962
https://github.com/nuxt/ui/issues/3377
https://github.com/nuxt/ui/issues/3597
https://github.com/nuxt/ui/issues/3821

NB: typecheck issue was already present before this commit
../ui/src/runtime/components/Table.vue:228:3 - error TS2322: Type 'Ref<UnwrapRefSimple<T>[], T[] | UnwrapRefSimple<T>[]>' is not assignable to type 'MaybeRef<T[]>'.
  Type 'Ref<UnwrapRefSimple<T>[], T[] | UnwrapRefSimple<T>[]>' is not assignable to type 'Ref<T[], T[]>'.
    Type 'UnwrapRefSimple<T>[]' is not assignable to type 'T[]'.
      Type 'UnwrapRefSimple<T>' is not assignable to type 'T'.
        'UnwrapRefSimple<T>' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint 'unknown'.
          Type 'unknown' is not assignable to type 'T'.
            'unknown' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint 'unknown'.

zguig52 committed:
Merge branch 'nuxt:v3' into dev

zguig52 committed:
fix(Table): ensure typescript checks are ok

zguig52 committed:
Merge branch 'dev' of github.com:zguig52/ui into dev

benjamincanac commented on test/components/Table.spec.ts:
```diff
@@ -149,6 +149,7 @@ describe('Table', () => {
     ['with caption', { props: { ...props, caption: 'Table caption' } }],
     ['with columns', { props: { ...props, columns } }],
     ['with sticky', { props: { ...props, sticky: true } }],
+    ['with disabled watchDeep', { props: { ...props, watchDeep: false } }],
```
I don't think this is necessary, the output won't change.

benjamincanac commented on src/runtime/components/Table.vue:
```diff
@@ -97,6 +97,10 @@ export interface TableProps<T extends TableData> extends TableOptions<T> {
+  /**
+   * @defaultValue true
+   */
+  watchDeep?: WatchOptions['deep']
```
Maybe we could allow the entire `WatchOptions` as a `watch` prop object? 🤔

benjamincanac commented on docs/content/3.components/table.md:
```diff
@@ -213,6 +213,9 @@ props:
   class: 'flex-1 max-h-[312px]'
 ---
 ::
+### WatchDeep
```
I don't think it's necessary to add this section, the prop will be documentation in the `### Props` anyway.

zguig52 committed:
fix(Table): Review requests changes implemented
- remove useless new test
- change watchDeep prop to more general watchOptions
- integrate documentation directly in table props description

benjamincanac committed:
up

benjamincanac commented:
Thanks @zguig52! 😊
